### DISCUSSION TITLE
Accessing and blocking specific line item discounts from third-party apps in Cart and Checkout Validation API

### DISCUSSION THREAD

**Discussion by basilkhan05** [Shopify/function-examples#231] [Feedback: Cart and Checkout Validation API]
Hello Shopify team and community!

I've received a request for a specific use case related to the cart and checkout validation functions concerning discount applications on line items from third-party apps (not using functions). This functionality could be beneficial for other merchants in our community as well. I'd like to provide some context and details on the issue and gather feedback on the potential for its implementation.

**Problem Statement**

A Shopify Plus merchant has inquired about the possibility of exposing discounts applied on a product or order through the cart and checkout validation input query. This would allow them to prevent specific discount codes from being applied via third-party apps (e.g. one-off birthday discounts for customers) when price-reduced products (using compare-at-price) are in the cart, such as bundles.

E.g. Bundle A is priced at $20 with compare-at-price at $24. However, a third-party app (SMSBump) gives 10% off on birthdays which adds another 10% off the order and hence applies the additional discount on the bundle. Our aim is to use cart/checkout validations to raise an error if additional discounts are applied to a specific line item.

**Current Limitations**

Cart and Checkout validations do not expose discount applications for either the order or line items in the cart.

**Proposed Solution**

Implement functionality in the Cart and Checkout Validation API to access and block specific discounts from third-party apps. This would enable merchants to have more control over the discounts applied to their products, particularly in situations where they want to prevent certain discounts from being applied to price-reduced items or bundles.

**Questions**

1. Are there any existing workarounds or alternative solutions for this use case that I may have missed?
2. Is there any plan to expose discounts applied on a product or order in the cart and checkout validation input query in the future?

Thank you for your time and consideration. I look forward to your feedback and any suggestions on how to address this use case effectively.

**Comment by aledalgrande** [comment]
Hey, thanks for going through the details of your use case.

Not sure if I am understanding correctly, but this seems a discount concern, not a validation concern. If you implemented a validation function like in your example, the buyer would get stuck because the discounts app is doing something unexpected. What we want is instead for the buyer to be able to checkout with no friction and with the right discounts applied to the cart.

From my point of view, it would be ideal if the app applying the discounts would check for price-reduced products first, and only then apply the discount if needed. Does that make sense?

**Reply by basilkhan05** [reply]
@aledalgrande Thanks for your reply!

> From my point of view, it would be ideal if the app applying the discounts would check for price-reduced products first, and only then apply the discount if needed. Does that make sense?

Yes, that makes sense. However, the use case here is the App # 1 manages products (bundles) and App # 2 manages discounts via SMS, not via the discounts function. Since App # 1 already reduces the price and bundles products (compare-at-price + price adjustment) to put the product on sale, the merchant wants customers to have the promotion/birthday discount code removed so the product isn't discounted even more than initially anticipated by both apps. Hope this makes sense. The bundle management app doesn't have access to App # 2 or the discounts applied at checkout via discount code links. Hence this question isn't related to the discount functions at all and is more around validating checkout so certain products aren't heavily discounted. 

> check for price-reduced products first

How would we check for `price-reduced products` in the cart validation function? 

Thanks! 🙏

**Reply by aledalgrande** [reply]
Hey, currently validation functions don't have access to discounts, so they wouldn't be able to validate like in this flow.
